### PR TITLE
Mdbenjam prerender docs

### PR DESCRIPTION
**Pull request by mdbenjam**
Connects: #1984 

The goal here is to make switching between documents faster. We accomplish this by prerendering documents on hidden DOM elements. We render the first two pages of the previous and next document. Once rendered, PDFJS make future renders much faster.

I also took out the debounce that prevented moving between documents quickly, and tried to root out any remaining rendering bugs.

Some numbers, all on the current PR, but looking at pages rendering before and after they'd been prerendered.

| Time to render on first render | Time to render after prerendered |
|--------------------------------|----------------------------------|
| 627.27 ms                      | 168.78 ms                          |
| 631.82 ms                      | 177.52 ms                          |

**Test Plan**
- [ ] Verify moving between documents is near instantaneous once as long as you aren't switching "too fast"
- [ ] Verify moving between documents quickly doesn't throw any errors, or show any visual artifacts other than seeing constant switching between documents.

### CODE REVIEW

File: client/app/components/Pdf.jsx [hunk 1]
```diff
@@ -1,3 +1,5 @@
+/* eslint-disable max-lines */
```
Comment (mdbenjam) [inline client/app/components/Pdf.jsx]: My plan is to systematically push every single file over the line limit.

Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: lololol

**Comment by NickHeiner** [comment]
When I go to http://localhost:3000/reader/appeal/reader_id1/documents/5, I get the following error:

image

**Comment by NickHeiner** [comment]
It would be awesome if we could get some numbers around how much of a difference this makes, but no need to do that if it's too much of a pain.

**Comment by NickHeiner** [comment]
Anecdotally, this does appear to speed up doc render time.

File: client/app/components/Pdf.jsx [hunk 2]
```diff
@@ -90,17 +96,21 @@ export class Pdf extends React.Component {
     this.isRendering[index] = true;
 
     return new Promise((resolve, reject) => {
-      if (index > this.state.numPages || pdfDocument !== this.state.pdfDocument) {
+      if (index >= this.state.numPages || pdfDocument !== this.state.pdfDocument) {
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: What fix are we making here and on the next line?

File: client/app/components/Pdf.jsx [hunk 3]
```diff
@@ -152,14 +162,26 @@ export class Pdf extends React.Component {
         // in the current state. It is possible that the
         // user switched between PDFs quickly and this
         // condition is no longer true, in which case we
-        // should render this page again with the new file.
-        if (pdfDocument === this.state.pdfDocument) {
+        // should render this page again with the new file. We
+        // also check if the canvas rendered on still exists.
+        // If the pages are changed quickly it's possible to
+        // render on a canvas that has since been change which
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: "since been changed" instead of "since been change"?

File: client/app/components/Pdf.jsx [hunk 4]
```diff
@@ -152,14 +162,26 @@ export class Pdf extends React.Component {
...
+        // means we need to render it again.
+        if (pdfDocument === this.state.pdfDocument && canvas === this.pageElements[index].canvas) {
           // If it is the same, then we mark this page as rendered
           this.setIsRendered(index, {
             pdfDocument,
             scale,
             ..._.pick(viewport, ['width', 'height'])
           });
+
+          // Whenever we finish rendering a page, we assume that we have some spare cycles
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: Can you explain more what you mean by this?

File: client/app/components/Pdf.jsx [hunk 5]
```diff
@@ -214,9 +237,9 @@ export class Pdf extends React.Component {
   }
 
   performFunctionOnEachPage = (func) => {
-    Array.prototype.forEach.call(this.pageContainers, (ele, index) => {
-      if (ele) {
-        const boundingRect = ele.getBoundingClientRect();
+    Array.prototype.forEach.call(this.pageElements, (ele, index) => {
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: No need for the `Array.prototype` call here. Just do `this.pageElements.forEach`.

File: client/app/components/Pdf.jsx [hunk 6]
```diff
@@ -214,9 +237,9 @@ export class Pdf extends React.Component {
   }
 
   performFunctionOnEachPage = (func) => {
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: This could also be:

```js
  performFunctionOnEachPage = (func) => 
    _(this.pageElements).
      filter('pageContainer').
      map((ele) => ele.pageContainer.getBoundingClientRect()).
      forEach(func)
```

Up to you if you'd like to change it.

File: client/app/components/Pdf.jsx [hunk 7]
```diff
@@ -225,24 +248,59 @@ export class Pdf extends React.Component {
 
   // This method sets up the PDF. It sends a web request for the file
   // and when it receives it, starts to render it.
-  setupPdf = _.debounce((file) => {
+  setupPdf = (file) => {
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: As long as this line is in the diff, I'll leave my usual nit about "setup" being a noun and "set up" being a verb, and functions being named with verbs, so this should be `setUpPdf`.

File: client/app/components/Pdf.jsx [hunk 8]
```diff
@@ -225,24 +248,59 @@ export class Pdf extends React.Component {
...
+    this.latestFile = file;
+
     return new Promise((resolve) => {
-      PDFJS.getDocument(file).then((pdfDocument) => {
+      this.getDocument(this.latestFile).then((pdfDocument) => {
+        // Don't continue seting up the pdf if it's already been setup.
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: "set up".

File: client/app/components/Pdf.jsx [hunk 9]
```diff
@@ -225,24 +248,59 @@ export class Pdf extends React.Component {
...
+
         this.setState({
           numPages: pdfDocument.pdfInfo.numPages,
           pdfDocument,
           isRendered: []
         }, () => {
-          resolve();
-          this.onPageChange(1);
+          // If the user moves between pages quickly we want to make sure that we just
+          // setup the most recent file. So we call this function recursively.
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: "set up"

Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: > So we call this function recursively.

Starting this sentence with "So" makes it a bit awkward. How about:

> set up the most recent file, so we call this function recursively.

or 

> set up the most recent file. Thus, we call this function recursively.

File: client/app/components/Pdf.jsx [hunk 10]
```diff
@@ -225,24 +248,59 @@ export class Pdf extends React.Component {
...
         this.props.setPdfReadyToShow(this.props.documentId);
       });
     });
-  }, DOCUMENT_DEBOUNCE_TIME, {
-    leading: true,
-    trailing: true
-  });
+  }
+
+  getDocument = (file) => {
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: This is more complicated than it needs to be.

```diff
diff --git a/client/app/components/Pdf.jsx b/client/app/components/Pdf.jsx
index 0d032046..daac902a 100644
--- a/client/app/components/Pdf.jsx
+++ b/client/app/components/Pdf.jsx
@@ -278,27 +278,22 @@ export class Pdf extends React.PureComponent {
   }
 
   getDocument = (file) => {
-    return new Promise((resolve, reject) => {
+    if (_.get(this.prerenderedPdfs, [file, 'pdfDocument'])) {
+      return Promise.resolve(this.prerenderedPdfs[file].pdfDocument);
+    }
+
+    return PDFJS.getDocument(file).then((pdfDocument) => {
+      // There is a chance another async call has resolved in the time that
+      // getDocument took to run. If so, again just use the cached version.
       if (_.get(this.prerenderedPdfs, [file, 'pdfDocument'])) {
-        resolve(this.prerenderedPdfs[file].pdfDocument);
-      } else {
-        PDFJS.getDocument(file).then((pdfDocument) => {
-          // There is a chance another async call has resolved in the time that
-          // getDocument took to run. If so, again just use the cached version.
-          if (_.get(this.prerenderedPdfs, [file, 'pdfDocument'])) {
-            resolve(this.prerenderedPdfs[file].pdfDocument);
-          } else {
-            this.prerenderedPdfs[file] = {
-              pdfDocument,
-              rendered: []
-            };
-            resolve(pdfDocument);
-          }
-        }).
-        catch(() => {
-          reject();
-        });
+        return this.prerenderedPdfs[file].pdfDocument;
       }
+      this.prerenderedPdfs[file] = {
+        pdfDocument,
+        rendered: []
+      };
+
+      return pdfDocument;
     });
   }
 

```

File: client/app/components/Pdf.jsx [hunk 11]
```diff
@@ -291,17 +351,69 @@ export class Pdf extends React.Component {
...
     }
     /* eslint-enable no-negated-condition */
   }
 
+  prerenderPages = () => {
+    // Don't prerender if we are currently trying to render a page on the current document.
+    // We want those pages to take precedence over pages on non-visible documents.
+    // At the end of rendering pages from this document we always call prerenderPages
+    // again in case there are still pages to prerender.
+    if (this.isRendering.some((value) => value)) {
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: This looks cleaner to me:

```js
if (_.some(this.isRendering)) {
```

File: client/app/components/Pdf.jsx [hunk 12]
```diff
@@ -291,17 +351,69 @@ export class Pdf extends React.Component {
...
+      return;
+    }
+
+    this.props.prefetchFiles.forEach((file, index) => {
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: The opening `{` on both this and the next line are unnecessary.

File: client/app/components/Pdf.jsx [hunk 13]
```diff
@@ -291,17 +351,69 @@ export class Pdf extends React.Component {
...
+      this.getDocument(file).then((pdfDocument) => {
+        _.range(NUM_PAGES_TO_PRERENDER).forEach((pageIndex) => {
+          if (pageIndex < pdfDocument.pdfInfo.numPages &&
+            !_.get(this.prerenderedPdfs, [file, 'rendered', pageIndex], false) &&
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: Why is the `false` necessary here? If the lookup doesn't match a path in the object, the default value of `undefined` is falsey anyway.

File: client/app/components/Pdf.jsx [hunk 14]
```diff
@@ -291,17 +351,69 @@ export class Pdf extends React.Component {
...
+            this.fakeCanvas[index][pageIndex] &&
+            !this.isPrerendering) {
+            // We set this to true, so that only one page can prerender at a time. In this
+            // way we can prerender page 1 before prerendering page 2.
+            this.isPrerendering = true;
+
+            this.prerenderedPdfs[file].rendered[pageIndex] = true;
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: Shouldn't we set this after the render actually succeeds?

File: client/app/components/Pdf.jsx [hunk 15]
```diff
@@ -291,17 +351,69 @@ export class Pdf extends React.Component {
...
+
+            pdfDocument.getPage(pageIndex + 1).then((pdfPage) => {
+              const viewport = pdfPage.getViewport(this.props.scale);
+
+              pdfPage.render({
+                canvasContext: this.fakeCanvas[index][pageIndex].getContext('2d', { alpha: false }),
+                viewport
+              }).
+              then(() => {
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: This function is repeated three times. Let's factor it out.

File: client/app/components/Pdf.jsx [hunk 16]
```diff
@@ -416,21 +526,44 @@ export class Pdf extends React.Component {
...
         id={`pageContainer${pageNumber}`}
         ref={(pageContainer) => {
-          this.pageContainers[pageNumber - 1] = pageContainer;
+          _.set(this.pageElements, [pageNumber - 1, 'pageContainer'], pageContainer);
         }}>
           <div className={pageContentsVisibleClass}>
             <canvas
-              id={`canvas${pageNumber}`}
+              id={`canvas${pageNumber}-${this.props.file}`}
+              ref={(canvas) => {
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: image

The render police will arrest you for this. Please never pass an inline function as a prop.

Also, no need for the `{` on this line.

File: client/app/components/Pdf.jsx [hunk 17]
```diff
@@ -416,21 +526,44 @@ export class Pdf extends React.Component {
...
+
+      const canvases = _.range(NUM_PAGES_TO_PRERENDER).map((pageIndex) =>
+        <canvas
+          style={{ display: 'none' }}
+          ref={(ele) => {
+            this.fakeCanvas[index][pageIndex] = ele;
+          }}/>
+      );
+
+      return <div>
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: I do not believe this `div` is necessary; I think we can `return` the result of the mapping on line 551.

File: client/app/reader/PdfViewer.jsx [hunk 18]
```diff
@@ -72,6 +72,13 @@ export class PdfViewer extends React.Component {
   prevDocId = () => _.get(this.props.documents, [this.selectedDocIndex() - 1, 'id'])
   nextDocId = () => _.get(this.props.documents, [this.selectedDocIndex() + 1, 'id'])
 
+  prefetchFiles = () => {
```
Comment (NickHeiner) [inline client/app/reader/PdfViewer.jsx]: Functions should be named with verbs, so let's make this `getPrefetchFiles`.

Comment (NickHeiner) [inline client/app/reader/PdfViewer.jsx]: This function could also be:

```js
  getPrefetchFiles = () => _.compact([
    this.prevDocId(),
    this.nextDocId() 
  ]).map(documentPath)
```

Comment (NickHeiner) [inline client/app/reader/PdfViewer.jsx]: All of the logic for this comes out of Redux, so there's no real point to having to pass this value all the way down from `PdfViewer`. But that would be a larger refactor, so this is fine for now.

File: client/app/components/Pdf.jsx (hunk 16 above)
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: Also please fix this below.

File: client/app/components/Pdf.jsx (hunk 2 above)
Comment (mdbenjam) [inline client/app/components/Pdf.jsx]: So if you quickly flip from one document with 4 pages to another document with 3 pages some bad things can happen.

Consider page 4 is told to render. Then we land on the 3 page document. Then this promise begins running. We check if the index of page 4 (3) is greater than this.state.numPages (3). Since 3 is not greater than 3 we will continue trying to render this and will get null for the values in canvas, container, and page on lines 107-109.

Furthermore if we resolve from this method, we need to reset isRendering or else this page index cannot render for other documents.

File: client/app/components/Pdf.jsx (hunk 6 above)
Comment (mdbenjam) [inline client/app/components/Pdf.jsx]: I generally find chained lodash calls harder to read. So I'm going to stick with the current version.

**Comment by NickHeiner** [comment]
Thanks for adding the perf numbers. How did you get them?

File: client/app/components/Pdf.jsx [hunk 19]
```diff
@@ -152,14 +163,30 @@ export class Pdf extends React.Component {
...
+        // means we need to render it again.
+        if (pdfDocument === this.state.pdfDocument && canvas === this.pageElements[index].canvas) {
           // If it is the same, then we mark this page as rendered
           this.setIsRendered(index, {
             pdfDocument,
             scale,
             ..._.pick(viewport, ['width', 'height'])
           });
+
+          // Whenever we finish rendering a page, we assume that we have some spare time.
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: Ha, sorry I was unclear. I know what "spare cycles" means; I just wasn't sure why we'd make this assumption. Sync chat?

File: client/app/components/Pdf.jsx [hunk 20]
```diff
@@ -225,24 +253,54 @@ export class Pdf extends React.Component {
...
-          resolve();
-          this.onPageChange(1);
+          // If the user moves between pages quickly we want to make sure that we just
+          // set up the most recent file, so we call this function recursively.
+          this.setUpPdf(this.latestFile).then(() => {
+            this.onPageChange(1);
+            resolve();
+          });
         });
         this.props.setPdfReadyToShow(this.props.documentId);
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: Is it odd that we're calling this before `setState` is done?

File: client/app/components/Pdf.jsx [hunk 21]
```diff
@@ -344,6 +453,14 @@ export class Pdf extends React.Component {
     event.preventDefault();
   }
 
+  getCanvasRef = (pageNumber) => (canvas) => _.set(this.pageElements, [pageNumber - 1, 'canvas'], canvas)
+  getTextRef = (pageNumber) => (textLayer) => _.set(this.pageElements, [pageNumber - 1, 'textLayer'], textLayer)
+  getFakeCanvasRef = (index, pageIndex) => (ele) => this.fakeCanvas[index][pageIndex] = ele
+  getPageContainerRef = (pageNumber) => (pageContainer) =>
+    _.set(this.pageElements, [pageNumber - 1, 'pageContainer'], pageContainer)
+  getScrollWindowRef = (scrollWindow) => this.scrollWindow = scrollWindow
+
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: nit: extra space

File: client/app/components/Pdf.jsx [hunk 22]
```diff
@@ -286,22 +346,71 @@ export class Pdf extends React.Component {
...
+            this.fakeCanvas[index][pageIndex] &&
+            !this.isPrerendering) {
+            // We set this to true, so that only one page can prerender at a time. In this
+            // way we can prerender page 1 before prerendering page 2.
+            this.isPrerendering = true;
+
+            pdfDocument.getPage(pageIndex + 1).then((pdfPage) => {
+              const viewport = pdfPage.getViewport(this.props.scale);
+
+              pdfPage.render({
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: This function can also be cleaned up to reduce nesting and duplication:

```diff
diff --git a/client/app/components/Pdf.jsx b/client/app/components/Pdf.jsx
index 8204574c..0f3c906b 100644
--- a/client/app/components/Pdf.jsx
+++ b/client/app/components/Pdf.jsx
@@ -390,15 +390,14 @@ export class Pdf extends React.PureComponent {
             pdfDocument.getPage(pageIndex + 1).then((pdfPage) => {
               const viewport = pdfPage.getViewport(this.props.scale);
 
-              pdfPage.render({
+              return pdfPage.render({
                 canvasContext: this.fakeCanvas[index][pageIndex].getContext('2d', { alpha: false }),
                 viewport
-              }).
-              then(() => {
-                this.prerenderedPdfs[file].rendered[pageIndex] = true;
-                finishPrerender();
-              }).
-              catch(finishPrerender);
+              });
+            }).
+            then(() => {
+              this.prerenderedPdfs[file].rendered[pageIndex] = true;
+              finishPrerender();
             }).
             catch(finishPrerender);
           }

```

File: client/app/components/Pdf.jsx [hunk 23]
```diff
@@ -415,31 +530,42 @@ export class Pdf extends React.Component {
...
         id={`pageContainer${pageNumber}`}
-        ref={(pageContainer) => {
-          this.pageContainers[pageNumber - 1] = pageContainer;
-        }}>
+        ref={this.getPageContainerRef(pageNumber)}>
           <div className={pageContentsVisibleClass}>
             <canvas
-              id={`canvas${pageNumber}`}
+              id={`canvas${pageNumber}-${this.props.file}`}
+              ref={this.getCanvasRef(pageNumber)}
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: image

This is still problematic, because we are still defining a new function in each `render` call.

```js
const fn = (a) => (b) => a + b

fn(1) !== fn(1)
```

Let me know if you'd like help restructuring this.

File: client/app/components/Pdf.jsx [hunk 24]
```diff
@@ -344,6 +453,14 @@ export class Pdf extends React.Component {
     event.preventDefault();
   }
 
+  getCanvasRef = (pageNumber) => (canvas) => _.set(this.pageElements, [pageNumber - 1, 'canvas'], canvas)
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: If we have functions that produce other functions, I'd prefer they not be named like normal getters. How about `makeGetCanvasRef`?

File: client/app/components/Pdf.jsx [hunk 25]
```diff
@@ -415,31 +530,42 @@ export class Pdf extends React.Component {
...
               className="textLayer"/>
           </div>
         </div>);
     }
-    this.scrollWindow = null;
+
+    const prerenderCanvases = this.props.prefetchFiles.map((_unused, index) => {
+      this.fakeCanvas[index] = [];
+
+      const canvases = _.range(NUM_PAGES_TO_PRERENDER).map((pageIndex) =>
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: Let's just `return` this directly.

File: client/app/components/Pdf.jsx [hunk 26]
```diff
@@ -415,31 +530,42 @@ export class Pdf extends React.Component {
...
               id={`textLayer${pageNumber}`}
+              ref={this.getTextRef(pageNumber)}
               className="textLayer"/>
           </div>
         </div>);
     }
-    this.scrollWindow = null;
+
+    const prerenderCanvases = this.props.prefetchFiles.map((_unused, index) => {
+      this.fakeCanvas[index] = [];
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: It feels odd to have a side-effect in a `map`. Can we do this in the `getRef` function instead?

**Comment by NickHeiner** [comment]
:+1: on the updated comments.

File: client/app/components/Pdf.jsx [hunk 27]
```diff
@@ -56,7 +58,15 @@ export class Pdf extends React.Component {
...
     this.isRendering = [];
-    this.pageContainers = [];
+    this.prerenderedPdfs = {};
+    this.isPrerendering = false;
+
+    this.pageElements = [];
+    this.fakeCanvas = [];
+    this.scrollWindow = null;
+
+    this.getRefFunctions = {};
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: Minor, but a variable that starts with `get` sounds like a getter to me. I would expect `getRefFunctions` to be a function that returned a collection of functions. What about `refFunctionGetters`? That sounds better to me at least.

File: client/app/components/Pdf.jsx [hunk 28]
```diff
@@ -225,24 +256,64 @@ export class Pdf extends React.Component {
...
+        }
+
+        this.pageElements = [];
+
+        this.getRefFunctions.canvas = [];
+        this.getRefFunctions.textLayer = [];
+        this.getRefFunctions.pageContainer = [];
+
+        _.range(pdfDocument.pdfInfo.numPages).forEach((index) => {
+          this.getRefFunctions.canvas[index] = this.makeGetCanvasRef(index);
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: At this point, I think it would be more readable just to inline these functions.

```js
this.getRefFunctions.canvas[index] = (canvas) => _.set(this.pageElements, [pageNumber, 'canvas'], canvas)
```

File: client/app/components/Pdf.jsx [hunk 29]
```diff
@@ -225,24 +256,64 @@ export class Pdf extends React.Component {
...
+          return resolve();
+        }
+
+        this.pageElements = [];
+
+        this.getRefFunctions.canvas = [];
+        this.getRefFunctions.textLayer = [];
+        this.getRefFunctions.pageContainer = [];
+
+        _.range(pdfDocument.pdfInfo.numPages).forEach((index) => {
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: What if we flipped the `index` and the `canvas|textLayer|pageContainer` key, so we could set up this object a bit more simply?

```js
        this.getRefFunctions = _.range(pdfDocument.pdfInfo.numPages).map((index) => ({
          canvas: canvasFn,
          textLayer: textLayerFn,
          pageContainer: pageContainerFn
        })
```

File: client/app/components/Pdf.jsx [hunk 30]
```diff
@@ -415,31 +556,38 @@ export class Pdf extends React.Component {
...
         id={`pageContainer${pageNumber}`}
-        ref={(pageContainer) => {
-          this.pageContainers[pageNumber - 1] = pageContainer;
-        }}>
+        ref={this.getRefFunctions.pageContainer[pageNumber - 1]}>
           <div className={pageContentsVisibleClass}>
             <canvas
-              id={`canvas${pageNumber}`}
+              id={`canvas${pageNumber}-${this.props.file}`}
+              ref={this.getRefFunctions.canvas[pageNumber - 1]}
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: It's annoying to have to go back and forth between 1-indexed page numbers and 0-indexed indexes. We could improve readability with `pageNumberOfIndex()` and `indexOfPageNumber()` functions, but we don't need to make that change now.

File: client/app/components/Pdf.jsx [hunk 31]
```diff
@@ -225,24 +256,64 @@ export class Pdf extends React.Component {
 
   // This method sets up the PDF. It sends a web request for the file
   // and when it receives it, starts to render it.
-  setupPdf = _.debounce((file) => {
+  setUpPdf = (file) => {
+    this.latestFile = file;
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: Why is this necessary?

File: client/app/components/Pdf.jsx (hunk 6 above)
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: fine

File: client/app/components/Pdf.jsx (hunk 29 above)
Comment (mdbenjam) [inline client/app/components/Pdf.jsx]: While that's nice, we're also using this object to hold the getRefFunction for the fakeCanvases, which don't have the same array structure.

File: client/app/components/Pdf.jsx (hunk 30 above)
Comment (mdbenjam) [inline client/app/components/Pdf.jsx]: Agreed on both counts.

File: client/app/components/Pdf.jsx (hunk 29 above)
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: In that case we could do:

```ts
{
    fakeCanvases: {},
    perPage: {
        [index: number]: {canvas: Function, textLayer: Function, pageContainer: Function}
    }
}
```

but w/e.

Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: Or just split them out into two vars entirely.
